### PR TITLE
Classes

### PR DESCRIPTION
**Pull request by armengau**
- Restructured viewer.py with classes (ViewerCDS, ViewerPlots, Viewer[VI]Widgets, ViewerSetup).
- Code tested both in test notebooks and with scripts generating html.

Ready to merge on my side.

### CODE REVIEW

**Comment by weaverba137** [comment]
I'm taking a look, but one initial comment.  Why choose to make `viewer_cds.py`, `viewer_plots.py`, `viewer_setup.py`... when you could put those all into a subpackage: `viewer/cds.py`, `viewer/plots.py`, `viewer/setup.py`?

I also updated the api documentation.

**Comment by armengau** [comment]
> I'm taking a look, but one initial comment. Why choose to make `viewer_cds.py`, `viewer_plots.py`, `viewer_setup.py`... when you could put those all into a subpackage: `viewer/cds.py`, `viewer/plots.py`, `viewer/setup.py`?

The rationale is: each file has a name close to its associated class eg ViewerCDS, ViewerPlot etc.

I'd rather avoid class names "CDS", "Plot", etc, such names would be too generic, too easy to be confused with.

On the other hand, I fully agree to have eg class ViewerCDS implemented in viewer/cds.py, if you think it's better.

**Comment by weaverba137** [comment]
Indeed.  When you use names like `viewer_plots`, you're just reinventing the wheel on top of the namespace facility Python already provides via `viewer/plots`.  However, you should still name the classes *e.g.* `ViewerCDS`.

**Comment by weaverba137** [comment]
In `prospect/scripts/specview_per_expo.py`:
```python
from ..utilities import match_zbest_to_spectra, match_vi_targets  #, miniplot_spectrum
```
There is no such function `match_zbest_to_spectra()` in `prospect.utilities`.

File: py/prospect/viewer_setup.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        for i_spec in range(nspec) :
+            self.thumb_callback = CustomJS(args=dict(full_viewer=self.full_viewer, i_spec=i_spec, ifiberslider=widgets.ifiberslider), code="""
+            full_viewer.active = 0
+             ifiberslider.value = i_spec
+            """)
+            (self.thumb_grid.children[i_spec][0]).js_on_event(bokeh.events.DoubleTap, self.thumb_callback)
+
+
+
+class ThumbGrid(object):
```
Comment (weaverba137) [inline py/prospect/viewer_setup.py:165]: Shouldn't this class be in grid_thumbs.py?

**Review by weaverba137 — COMMENTED** [review]
I found an apparently misplaced class, but overall I think this looks fine.  We just need to do a full test with data on various systems.  I should be able to test on Data Lab with SDSS data later this week.  There are some problems with that system right now.

**Comment by weaverba137** [comment]
PS, Let's *not* attempt the rearrangement `viewer_cds.py` --> `viewer/cds.py` right now.  Let's concentrate on testing the branch as-is, and then if everything works we can rearrange or save that for another PR.

**Comment by armengau** [comment]
> In `prospect/scripts/specview_per_expo.py`:
> 
> ```python
> from ..utilities import match_zbest_to_spectra, match_vi_targets  #, miniplot_spectrum
> ```
> 
> There is no such function `match_zbest_to_spectra()` in `prospect.utilities`.

Indeed. This script specview_per_expo.py is outdated since early cmx - but I won't update it in this branch.
Fixed in https://github.com/desihub/prospect/pull/67/commits/695181cac0549057340e78200aa50cda2ae3efe6

**Comment by weaverba137** [comment]
OK, let's move to testing on DESI & SDSS data next.

File: py/prospect/viewer_setup.py (hunk 1 above)
Comment (armengau) [reply py/prospect/viewer_setup.py]: You're right the name is very misleading. I changed the class name in https://github.com/desihub/prospect/pull/67/commits/ecb68de7effbddd0a852255df6bdda64df7ceb02

=> Class `StandaloneThumbSetup`
The class does not implement the thumb grid, but rather the bokeh setup used to display this grid without the main prospect viewer. It has the same function as `ViewerSetup`, although much, much lighter.

Comment (weaverba137) [reply py/prospect/viewer_setup.py]: Good, thank you.

**Comment by armengau** [comment]
I redid the tests: 
- on a small test notebook
- running the same script as to create a subset of blanc VI pages (https://data.desi.lbl.gov/desi/spectro/prospect/dev/test/tiles_QSO/80605/deep/)
- on two notebooks in doc/nb, which I just updated (Prospect_demo.ipynb and Prospect_targetid.ipynb)

Looks ok for me.

If you agree I would prefer to rearrange the dir tree, eg `viewer_cds` --> `viewer/cds`, in this PR. Btw, where would you locate viewer.py ?

**Comment by weaverba137** [comment]
`viewer.py` can become `viewer/__init__.py`.  There are slightly more complicated things you can try, but if you want to test the rearrangement, start with that.

I'll test on Data Lab as soon as I can.

**Comment by armengau** [comment]
Last commits https://github.com/desihub/prospect/pull/67/commits/53b2c88126341e08f3f565c03329e254ec6e828c : changed the dir tree as suggested.
I also edited "Setup" classes to "Layouts", to avoid having a setup.py file...

Tested on notebooks+scripts, ok on my side.

**Comment by weaverba137** [comment]
I tested on Data Lab this morning and found no problems.  Merge when ready.  After merge I will create a new tag.

**Comment by weaverba137** [comment]
I also added an update to one notebook.

**Pull request merged**